src/mavsdk/core/mavlink_ftp_client.cpp

```diff
@@ -495,96 +498,88 @@ bool MavlinkFtpClient::download_burst_continue(
+            item.ofstream.write(empty.data(), empty.size());
+            if (!item.ofstream) {
+                LogWarn() << "Write failed";
+                item.callback(ClientResult::FileIoError, {});
+                download_burst_end(work);
+                return false;
+            }
         }
 
         // Write actual data to file.
         item.ofstream.write(reinterpret_cast<const char*>(payload->data), payload->size);
         if (!item.ofstream) {
             LogWarn() << "Write failed";
             item.callback(ClientResult::FileIoError, {});
+            download_burst_end(work);
             return false;
         }
 
         // Keep track of what was written.
-        for (size_t i = payload->offset; i < payload->offset + payload->size; ++i) {
-            item.transferred[i] = DownloadBurstItem::Transferred::Yes;
-        }
+        item.next_offset = payload->offset + payload->size;
 
         if (_debugging) {
             LogDebug() << "Received " << payload->offset << " to "
                        << payload->size + payload->offset;
         }
 
-        const size_t bytes_transferred = burst_bytes_transferred(item);
-
-        if (bytes_transferred == item.transferred.size()) {
+        if (payload->size + payload->offset >= item.file_size) {
             if (_debugging) {
                 LogDebug() << "Burst complete";
             }
 
-            // Final step
-            work.last_opcode = CMD_TERMINATE_SESSION;
-
-            work.payload = {};
-            work.payload.seq_number = work.last_sent_seq_number++;
-            work.payload.session = _session;
-
-            work.payload.opcode = work.last_opcode;
-            work.payload.offset = 0;
-            work.payload.size = 0;
-
-            start_timer();
-            send_mavlink_ftp_message(work.payload);
-
-            return true;
+            if (item.missing_data.empty()) {
+                // No missing data, we're done.
 
+                // Final step
+                download_burst_end(work);
+            } else {
+                // The burst is supposedly complete but we still need data because
+                // we missed some, so request next without burst.
+                request_next_rest(work, item);
+            }
         } else {
             item.callback(
                 ClientResult::Next,
                 ProgressData{
-                    static_cast<uint32_t>(bytes_transferred),
-                    static_cast<uint32_t>(item.transferred.size())});
+                    static_cast<uint32_t>(burst_bytes_transferred(item)),
+                    static_cast<uint32_t>(item.file_size)});
 
             if (payload->burst_complete) {
-                // The burst is supposedly complete but we still need data, so request next without
-                // burst.
+                // This burst is complete but the file isn't. we need to start a
+                // new one
                 request_next_rest(work, item);
```

Comment:
I think we can use burst download with offset (it works with PX4): https://mavlink.io/en/services/ftp.html#reading-a-file-burstreadfile
`Client (i.e. GCS) sends BurstReadFile command specifying the part of the file that it wants to get from an offset to the end, along with the default size of each burst payload.
The payload must specify: session: the current session id, offset = offset in file of start of burst, size= default length of payload in burst responses, data is ignored.`

I changed this because PX4 stop burst after 35KB and then download with standard read is slow (very slow when using serial line).
So we can start bursts until we don't receive the last data of the file and then ask for missing parts with standard reads.